doc/bgp_loading_optimization/bgp-loading-optimization-hld.md

```diff
@@ -0,0 +1,421 @@
+
+<figure align=center>
+    <img src="images/pipeline-timeline.png">
+    <figcaption>Figure 7. Pipeline architecture compared with the original serial architecture<figcaption>
+</figure>  
+
+#### Ring buffer for low-cost thread coordination
+Since multiple threads are employed, we take a lock-free design by using a ring buffer as an asynchronous communication channel.
+
+#### Asynchronous sairedis API usage
+Asynchronous mode `sairedis` API is used and a list of context of response pending messages is maintained on `orchagent` to process the response when its received
+
+<figure align=center>
+    <img src="images/async-sairedis2.png" width="40%" height=20%>
+    <figcaption>Figure 8. Async sairedis workflow<figcaption>
+</figure> 
+
+#### New ResponseThread in OA
+A new `ResponseThread` is used in `orchagent` to process the response when its received so that the other threads can continue processing new routing messages
+
+### Streamlining Redis I/O
+
+The optimization for `orchagent` and `syncd` can theoretically double the BGP loading performance, which makes Redis I/O performance become a new bottleneck.
+
+#### Lower frequency of the fpmsyncd flush & APPL_DB publish
+
+Instead of flushing the `pipeline` on every data arrival and propose to use a flush timer to determine the flush frequency as illustrated below.
+
+<figure align=center>
+    <img src="images/pipeline-mode.png" height=auto>
+    <figcaption>Figure 9. Proposed new BGP loading workflow<figcaption>
+</figure>  
+
+#### Disable the temporary table mechanism in APPL_DB
+
+We propose to disable the temporary/stable table behavior and keep just a single table, so that we don't need to delete the temporary and then write into a stable one, which spares much `HDEL` and `HSET` traffic.
+
+## Low-Level Implementation
+
+### Multi-threaded orchagent with a ring buffer
+
+Orchagent now runs two threads in parallel instead of a single thread.
+
+- `table->pops(entries)` executes in the master thread to maintain the time sequence
+- `Consumer->drain()` runs in a slave thread
+- `Consumer->addToSync(entries)` is run by slave, as master is assumed to be busier
+- `RingBuffer` is used for communication between the master thread and the slave
+  - the master thread pops  `entries` to the ring buffer
+  - the slave thread fetches `entries` from the ring buffer
+  
+Since SAI doesn't work well on small piece of data, the slave thread should check data size in ring buffer before it calls `Consumer->addToSync(entries)` to fetch data from the ring buffer, hence ensuring that it gets large enough data.
+
+Routes will still be cached in `Consumer->m_toSync` rather than ring buffer if routeorch fails to push route to ASIC_DB. 
+
+We use a new C++ class `Consumer_pipeline`, which is derived from the original `Consumer` class in `RouteOrch`, which enables the usage of a slave thread and utilizes the ring buffer.
```

Comment:
>Consumer_pipeline [](http://example.com/codeflow?start=24&length=17)

Suggest name (and naming convention): RingBufferConsumer. We already use "pipeline" for RedisPipeline in codebase, so better to be less confusing.